ansible/roles/azure/tasks/create_adlsgen2.yml

```diff
@@ -0,0 +1,235 @@
+
+- name: Retrieve sequence end number to get the number of storage accounts 
+  set_fact:
+     InstanceVolumesEndSequence: "{{ '1' if  instance_volumes_input.split('|')[0].split(',') == ['']  else InstanceVolumesTemp[0]|int }}"
+
+- name: Generate names for Storage Accounts
+  set_fact:
+     InstanceVolumesAuto: "{{ InstanceVolumesAuto + ['abfss://'+'accumulodata'+'@'+StorageAccountName+item+'.'+InstanceVolumesTemp[1]+'/accumulo'] }}"
+  with_sequence: start=1 end={{ InstanceVolumesEndSequence|int }}
+  when: InstanceVolumesTemp[0]|int != 0
+
+- name: Retrieve ABFSS values when specified manually 
+  set_fact:
+     InstanceVolumesManual: "{{ InstanceVolumesManual +  [ item ] }}"
+  loop:
+    "{{ InstanceVolumesTemp }}"
+  when: item.split('://')[0] == 'abfss' and  instance_volumes_input.split('|')[0].split(',') ==  ['']
+
+# This is  final list of instance volumes 
+- name: Assign variables for autogeneration or manual for storage account creation
+  set_fact:
+     InstanceVolumes: "{{ InstanceVolumesManual if  instance_volumes_input.split('|')[0].split(',') ==  [''] else InstanceVolumesAuto }}" 
+
+- name: Update instance_volumes_preferred  in muchos.props
+  lineinfile:
+    path: "{{ deploy_path }}/conf/muchos.props"
+    regexp: '^instance_volumes_preferred\s*=\s*|^[#]instance_volumes_preferred\s*=\s*'
+    line: "instance_volumes_preferred = {{ InstanceVolumes|join(',') }}"
+
+# Not registering variable because  storage values are not visible immediately
+- name: Create ADLS Gen2 storage acount using REST API
+  azure_rm_resource:
+    resource_group: "{{ resource_group }}"
+    provider: Storage
+    resource_type: storageAccounts
+    resource_name: "{{ item.split('@')[1].split('.')[0] }}"
+    api_version: '2019-04-01'
+    idempotency: yes
+    state: present
+    body:
+      sku:
+         name: "{{ adls_storage_type }}"
+      kind: StorageV2
+      properties:
+         isHnsEnabled: yes
+      location:  "{{ location }}"
+  loop:
+      "{{ InstanceVolumes }}"
+
+# Creating User Assigned identity with vmss_name suffixed by ua-msi if not specified in muchos.props
+# Not registering variable because user identity values are not visible immediately
+- name: Create User Assigned Identity 
+  azure_rm_resource:
+    resource_group: "{{ resource_group }}"
+    provider: ManagedIdentity
+    resource_type: userAssignedIdentities
+    resource_name: "{{ user_assigned_identity if user_assigned_identity !='' else vmss_name + '-ua-msi' }}"
+    api_version: '2018-11-30'
+    idempotency: yes
+    state: present
+    body:
+      location:  "{{ location }}"
+
+# Retrieving  facts about User Assigned Identity 
+- name: Get facts for User Assigned Identity
+  azure_rm_resource_facts:
+    resource_group: "{{ resource_group }}"
+    provider: ManagedIdentity
+    resource_type: userAssignedIdentities
+    resource_name: "{{ user_assigned_identity if user_assigned_identity !='' else vmss_name + '-ua-msi' }}"
+    api_version: '2018-11-30'
```

Comment:
Data Lake Storage Gen2 uses an access control model that supports both role-based access control (RBAC) and POSIX-like access control lists (ACLs).  Access to data in Data Lake Storage Gen2 is controlled through managed identities. A managed identity is an identity registered in Azure Active Directory (Azure AD) whose credentials are managed by Azure. With managed identities, you don't need to register service principals in Azure AD or maintain credentials such as certificates. Azure services have two types of managed identities: system-assigned and user-assigned.   For further information see https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/overview and https://hadoop.apache.org/docs/current/hadoop-azure/abfs.html#Authentication

Reply:
I was actually curious about `api_version`.  However I learned something from your explanation about managed identities.  Thanks for the explanation.

Reply:
For api_versiions, see https://docs.microsoft.com/en-us/rest/api/managedidentity/userassignedidentities/createorupdate

Hope I was able to help.  Otherwise please let me know

Reply:
Yes that does help.   I was wondering why there were different versions in this file (like `api_version: '2018-11-30'` and `api_version: '2019-04-01'`).  Looking around in the link that you sent, I am guessing that the API version goes with the provider (like `provider: Storage` or `provider: ManagedIdentity`)?  

Starting with that link, I found https://docs.microsoft.com/en-us/rest/api/storagerp/ and  I see it has a list of versions which  I assume are for the Storage provider.